### PR TITLE
[IMP] link_tracker: ignore requests from social bots for link tracker

### PR DESCRIPTION
**Pull request by fbarbe00**
Purpose
=======
Avoid counting requests from social bots (twitter, facebook, linkedin...) when tracking a link.

Specifications
=============
Social media platforms have a specific user agent in the HTTP headers that can be used to detect them and to not increment the click count in that case.

Task-2578902

### CODE REVIEW

File: addons/link_tracker/controller/main.py [hunk 1]
```diff
@@ -5,15 +5,23 @@
 from odoo.http import request
 
 
+ignore_user_agent_prefix = ["twitterbot", "facebookexternalhit", "facebot",
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:8]: use single quote for consistency :) (like we do in this file)

File: addons/link_tracker/controller/main.py [hunk 2]
```diff
@@ -5,15 +5,23 @@
...
+                            "LinkedInBot", "WhatsApp", "Googlebot", "Pinterest", "Instagram"]
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:9]: Your commit message can be improved
```
Add a manual list of user agent prefixes in odoo/addons/link_tracker/controller/main.py , then add an if statement on the route.
```

Here you describe what you did in your code, but I can see it by opening the commit.

A grand sage would say; the commit should not be really about "What", but more about "Why": "The social media need to build a preview, for that they make an HTTP call to the URLs in the posts. Those HTTP requests are currently count as user click". And "How": "Those social media have a specific user agent in the HTTP headers that can be used to detect them and to not increment the click count in that case".

Also update your PR message :)

File: addons/link_tracker/controller/main.py [hunk 3]
```diff
@@ -5,15 +5,23 @@
...
     def full_url_redirect(self, code, **post):
-        country_code = request.session.geoip and request.session.geoip.get('country_code') or False
-        request.env['link.tracker.click'].sudo().add_click(
-            code,
-            ip=request.httprequest.remote_addr,
-            country_code=country_code
-        )
+
+        user_agent = str(request.httprequest.user_agent) or ""
+        if not any(user_agent.lower().startswith(agent.lower()) for agent in ignore_user_agent_prefix):
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:18]: Can you change all `ignore_user_agent_prefix` in lower case, and just do `startswith(agent)` ?

File: addons/link_tracker/controller/main.py [hunk 4]
```diff
@@ -5,15 +5,23 @@
...
+            country_code = request.session.geoip and request.session.geoip.get(
+                'country_code') or False
+            request.env['link.tracker.click'].sudo().add_click(
+                code,
+                ip=request.httprequest.remote_addr,
+                country_code=country_code
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:24]: while you are at it, you can add a `,` :p

**Review by std-odoo — COMMENTED** [review]
Hi :)

Very good idea for this one, GG :) I already added some comment to speed-up a bit :)

File: addons/link_tracker/controller/main.py [hunk 5]
```diff
@@ -5,15 +5,23 @@
 from odoo.http import request
 
 
+ignore_user_agent_prefix = ['twitterbot', 'facebookexternalhit', 'facebot',
+                            'linkedinbot', 'whatsapp', 'googlebot', 'pinterest', 'instagram']
+
+
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:11]: the lines in your commit message are too long, generally we split at ~72 chars

File: addons/link_tracker/controller/main.py [hunk 6]
```diff
@@ -5,15 +5,23 @@
...
     @http.route('/r/<string:code>', type='http', auth='public', website=True)
     def full_url_redirect(self, code, **post):
-        country_code = request.session.geoip and request.session.geoip.get('country_code') or False
-        request.env['link.tracker.click'].sudo().add_click(
-            code,
-            ip=request.httprequest.remote_addr,
-            country_code=country_code
-        )
+
+        user_agent = str(request.httprequest.user_agent) or ""
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:17]:
```suggestion
        user_agent = str(request.httprequest.user_agent) or ''
```

Comment (std-odoo) [reply addons/link_tracker/controller/main.py:17]: why do you need to call `str` ?

File: addons/link_tracker/controller/main.py [hunk 7]
```diff
@@ -5,15 +5,23 @@
...
+        if not any(user_agent.lower().startswith(agent) for agent in ignore_user_agent_prefix):
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:18]: can you lower case the agent above ? so it be "lower cased" only once

**Review by std-odoo — COMMENTED** [review]
Hi :) 

I added some comments :)

File: addons/link_tracker/controller/main.py (hunk 6 above)
Comment (fbarbe00) [reply addons/link_tracker/controller/main.py]: because `request.httprequest.user_agent` is of class `<class 'werkzeug.useragents.UserAgent'>`, and we need to call .lower() on that.

Comment (fbarbe00) [reply addons/link_tracker/controller/main.py]: I changed it to `request.httprequest.user_agent.string`

Comment (std-odoo) [reply addons/link_tracker/controller/main.py]: so you do not need `or ''` anymore right ?

Comment (fbarbe00) [reply addons/link_tracker/controller/main.py]: Yep that works ! I was afraid that if no User Agent was sent it would be None and crash when comparing to a string, but turns out it's just an empty string when not set

File: addons/link_tracker/controller/main.py [hunk 8]
```diff
@@ -5,15 +5,22 @@
 from odoo.http import request
 
 
+ignore_user_agent = ['twitterbot', 'facebookexternalhit', 'facebot', 'applebot', 'bingbot', 'trendsmapresolver'
+                            'livelapbot', 'linkedinbot', 'whatsapp', 'googlebot', 'pinterest', 'instagram']
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:9]: indentation issue (are you sure your linter is correctly configured ?)
screenshot_spect

**Review by std-odoo — COMMENTED** [review]
Hi :) I had a look :)

Can you double check that flake8 is configured properly in your IDE ? :)

File: addons/link_tracker/controller/main.py [hunk 9]
```diff
@@ -9,11 +9,14 @@ class LinkTracker(http.Controller):
 
     @http.route('/r/<string:code>', type='http', auth='public', website=True)
     def full_url_redirect(self, code, **post):
-        country_code = request.session.geoip and request.session.geoip.get('country_code') or False
```
Comment (rdeodoo) [inline addons/link_tracker/controller/main.py:12]: You can't do that, http_routing is not a dependency of link_tracker, it will crash

File: addons/http_routing/models/ir_http.py [hunk 10]
```diff
@@ -331,17 +331,14 @@ def _get_translation_frontend_modules_name(cls):
...
     @classmethod
     def is_a_bot(cls):
         # We don't use regexp and ustr voluntarily
         # timeit has been done to check the optimum method
-        user_agent = request.httprequest.environ.get('HTTP_USER_AGENT', '').lower()
-        try:
-            return any(bot in user_agent for bot in cls.bots)
-        except UnicodeDecodeError:
-            return any(bot in user_agent.encode('ascii', 'ignore') for bot in cls.bots)
+        user_agent = request.httprequest.user_agent.string.lower()
```
Comment (rdeodoo) [inline addons/http_routing/models/ir_http.py:340]: Out of curiosity, why this change?
(and also about the try/catch)

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Hi :wave:

In the commit it says 
> User-agent is also added to link tracker view (hidden) to better identify bots.

But since you don't create the tracker if this is a bot, is this true?

**Comment by rdeodoo** [comment]
Side note: you can omit the PR reference in your commit message, it will be added to the commit by the mergebot.
It is especially true since very recently, where now multi-commit PR also got that behavior (previously it would only add it on the last commit of the PR).
It will prevent the "noise" on the PR :-)

image

Why didn't github yet provided a button/checkbox to hide those notification from a PR thread, who knows..

File: addons/link_tracker/controller/main.py (hunk 9 above)
Comment (fbarbe00) [reply addons/link_tracker/controller/main.py:12]: Hi rde, thanks for your fast review! I guess I should have marked it as draft, as I was still testing if the bot filter worked properly. I have moved the is_a_bot function in tools :).

File: addons/http_routing/models/ir_http.py (hunk 10 above)
Comment (fbarbe00) [reply addons/http_routing/models/ir_http.py]: changed ;)

File: odoo/tools/misc.py [hunk 11]
```diff
@@ -1641,3 +1641,14 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
         message.encode(),
         hash_function,
     ).hexdigest()
+
+
+bots = ["bot", "crawl", "slurp", "spider", "curl", "wget", "facebookexternalhit", "whatsapp", "trendsmapresolver", "pinterest", "instagram"]
+def is_a_bot(request):
```
Comment (std-odoo) [inline odoo/tools/misc.py:1647]: isn't it better to give `user_agent` directly ?

File: odoo/tools/misc.py [hunk 12]
```diff
@@ -1641,3 +1641,14 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
...
+    # We don't use regexp and ustr voluntarily
+    # timeit has been done to check the optimum method
```
Comment (std-odoo) [inline odoo/tools/misc.py:1649]: not very relevant comment

File: addons/link_tracker/controller/main.py [hunk 13]
```diff
@@ -3,17 +3,21 @@
...
+                'country_code') or False
+            request.env['link.tracker.click'].sudo().add_click(
+                code,
+                ip=request.httprequest.remote_addr,
+                user_agent=request.httprequest.user_agent,
+                country_code=country_code,
+            )
         redirect_url = request.env['link.tracker'].get_url_from_code(code)
-        return request.redirect(redirect_url or '', code=301, local=False)
+        return request.redirect(redirect_url or '/404.html', code=301, local=False)
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:23]: where is `/404.html` defined ? (do the same as "/qsdqsdqsd" so it's better to raise a NotFound error I think)

**Review by std-odoo — COMMENTED** [review]
Hi :) Some comments about the last changes you pushed :)

File: addons/link_tracker/controller/main.py (hunk 9 above)
Comment (rdeodoo) [reply addons/link_tracker/controller/main.py:12]: Oh, ok :-)
I figured since there were reviews going on and it was open, it was approaching shipping time :D 
Saw the mail notification about this pr since it changes http routing so I opened it out of curiosity :)

File: odoo/tools/misc.py (hunk 12 above)
Comment (rdeodoo) [reply odoo/tools/misc.py]: Isn't that comment relevant? It will prevent further change to revert what was done as one would be tempted to use a regex here

**Comment by rdeodoo** [comment]
> The issue is that the filter to check if it's a tracker will not always identify bots... I'll see if lna wants to keep this but this would make it much easier to identify bots that should not be tracked but are.

Thanks for the explanation
Makes sense, definitely worth a comment in the code before the if is_a_bot condition (and probably in the commit msg)

File: odoo/tools/misc.py (hunk 12 above)
Comment (std-odoo) [reply odoo/tools/misc.py]: I do not think, it's like each time you do `"str" in a` you said "we didn't use regex, or a library, or a C extension to do that because it's slower" (and it's obvious that regex will be slower...)

Comment (rdeodoo) [reply odoo/tools/misc.py]: I did Edit my message the same time you wrote yours :yum: 

It is indeed obvious, and most of the time it does not matter that much, that's why some colleagues will always prefer to change that kind of code to a regex when doing a quick fix or improvement on such code, which is fine most of the time.

But in this particular case, this method is called on every http request during the dispatching, where any gain is obviously very important.
That's why this comment seems relevant here :-)

Just my opinion tho.

File: addons/http_routing/models/ir_http.py [hunk 14]
```diff
@@ -389,7 +378,8 @@ def _add_dispatch_parameters(cls, func):
         if request.routing_iteration == 1:
             context = dict(request.context)
             path = request.httprequest.path.split('/')
-            is_a_bot = cls.is_a_bot()
+            user_agent = request.httprequest.user_agent.string.lower()
+            is_a_bot = is_a_bot(user_agent)
```
Comment (std-odoo) [inline addons/http_routing/models/ir_http.py:382]: the variable `is_a_bot` will overwrite the function `is_a_bot`, can you do `mics.is_a_bot` instead ?

File: addons/http_routing/models/ir_http.py [hunk 15]
```diff
@@ -466,7 +456,8 @@ def _dispatch(cls):
             path = request.httprequest.path.split('/')
             default_lg_id = cls._get_default_lang()
             if request.routing_iteration == 1:
-                is_a_bot = cls.is_a_bot()
+                user_agent = request.httprequest.user_agent.string.lower()
+                is_a_bot = is_a_bot(user_agent)
```
Comment (std-odoo) [inline addons/http_routing/models/ir_http.py:460]: same here

File: addons/link_tracker/controller/main.py [hunk 16]
```diff
@@ -3,17 +3,25 @@
 
 from odoo import http
 from odoo.http import request
+from odoo.tools.misc import is_a_bot
+from werkzeug.exceptions import NotFound
```
Comment (std-odoo) [inline addons/link_tracker/controller/main.py:7]: Sorry to be picky :/ werkzeug must be before odoo
https://www.odoo.com/documentation/15.0/developer/misc/other/guidelines.html#imports

File: addons/http_routing/models/ir_http.py [hunk 17]
```diff
@@ -389,7 +378,8 @@ def _add_dispatch_parameters(cls, func):
         if request.routing_iteration == 1:
             context = dict(request.context)
             path = request.httprequest.path.split('/')
-            is_a_bot = cls.is_a_bot()
+            user_agent = request.httprequest.user_agent.string.lower()
+            is_a_bot = misc.is_a_bot(user_agent)
```
Comment (odony) [inline addons/http_routing/models/ir_http.py:382]: if `misc.is_a_bot()` only works with lowercase UAs, it should do the lowercasing itself, instead of expecting all callers to guess that they need to do it ;)

**Review by odony — COMMENTED** [review]
PS: indeed, draft mode is a good idea if you want to avoid early reviews/noise ;)

File: odoo/tools/misc.py (hunk 12 above)
Comment (xmo-odoo) [reply odoo/tools/misc.py:1657]: > I do not think, it's like each time you do "str" in a you said "we didn't use regex, or a library, or a C extension to do that because it's slower" (and it's obvious that regex will be slower...)

It's not necessarily obvious that an alternate regex will be slower than an O(n^2) containment check though.

Comment (std-odoo) [reply odoo/tools/misc.py:1657]:
```
any(bot in user_agent for bot in cls.bots)
```
is O(n) not O(n^2)

File: odoo/tools/misc.py [hunk 18]
```diff
@@ -1641,3 +1641,21 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
...
+
+bots = ["bot", "crawl", "slurp", "spider", "curl", "wget", "facebookexternalhit", "whatsapp", "trendsmapresolver", "pinterest", "instagram"]
+def is_a_bot(user_agent):
+    """
+    Try to match if the given user agent is a bot or not based on a list of
+    bot identifiers.
+
+        :param str user_agent: user agent string to check
+        :return: True if the user agent is a bot, else False
+        """
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:1654]: Indentation is not correct.

File: odoo/tools/misc.py [hunk 19]
```diff
@@ -1641,3 +1641,21 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
         message.encode(),
         hash_function,
     ).hexdigest()
+
+
+bots = ["bot", "crawl", "slurp", "spider", "curl", "wget", "facebookexternalhit", "whatsapp", "trendsmapresolver", "pinterest", "instagram"]
+def is_a_bot(user_agent):
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:1647]: `odoo.tools.misc` seems like a pessimal place for this function, why not leave it where it was, or as a utlity function in `odoo.http`?

Converting it to a function also precludes overriding the method to add more filters, which was a feature of the previous location.

File: odoo/tools/misc.py [hunk 20]
```diff
@@ -1641,3 +1641,21 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
...
+        :param str user_agent: user agent string to check
+        :return: True if the user agent is a bot, else False
+        """
+    user_agent = user_agent.lower()
+    # We don't use regexp and ustr voluntarily
+    # timeit has been done to check the optimum method
+    try:
+        return any(bot in user_agent for bot in bots)
+    except UnicodeDecodeError:
+        return any(bot in user_agent.encode('ascii', 'ignore') for bot in bots)
```
Comment (xmo-odoo) [inline odoo/tools/misc.py:1661]: Fallback should be removed, either `user_agent` is actually `bytes` and should always be decoded, or `user_agent` is always `str` and the error will never occur. This is likely a remnant of Python 2 code.

**Review by xmo-odoo — CHANGES_REQUESTED** [review]

File: odoo/tools/misc.py (hunk 12 above)
Comment (xmo-odoo) [reply odoo/tools/misc.py:1657]: O(m*n) technically, it's the number of bots * the length of the UAS (UAS tend to be quite long).

Comment (JKE-be) [reply odoo/tools/misc.py:1657]: with this move, you cannot properly override the function :fearful:

Comment (xmo-odoo) [reply odoo/tools/misc.py]: @JKE-be noted the same issue in https://github.com/odoo/odoo/pull/78806/files/7372363bbc817bcc73e18622641e37d7f90fd4b4#diff-61162ac65633a1c7b054fc83ce1813f1a7984e3169ff36021713ef441f62a208

File: odoo/tools/misc.py (hunk 19 above)
Comment (fbarbe00) [reply odoo/tools/misc.py]: Hi,
Thanks for your comments. I moved the function in `odoo.http`. While it is true that the method can't be overwritten now, I can't think of a reason why this method would need to be overwritten.

The method is trying to identify bots based on user_agent, so if a new user agent for a bot is found, it should just be added here. It doesn't make much sense to me to have a request that is only seen as a bot in certain modules...

Let me know what you and @JKE-be think.

Comment (JKE-be) [reply odoo/tools/misc.py]: Because if you use a social network from an exotic country that Odoo don't have handled, you will be happy to be able to add it as a bot, instead to wait a long time that we never merge your PR in core code.

And because this move doesn't bring any advantage, only a move where you lose the opportunity to override it...

Comment (rdeodoo) [reply odoo/tools/misc.py]: This is an open source software, we provide core features that usually get extended by our community/partners to provide a layer on top of what we build, to fit their need, which to be honest, and it is normal, is not always sufficient to do what they need. (We can't support everything for everyone over the world)

They can't add code themselves in our codebase, so they extend this type of method, because the crawler in their country that no one knows is not in our codebase list of bot.

And I am pretty sure what they meant to keep it overidable was to keep it on ir.http model, not as a function.

Comment (fbarbe00) [reply odoo/tools/misc.py]: Alright, makes sense. I moved it to `odoo/addons/web/models/ir_http.py` after discussing it with std. Keeping it in `odoo/addons/http_routing/models/ir_http.py` wasn't really an option since `link_tracker` did not depend on it and adding a dependency didn't make much sense. Hope you agree with the current solution :)

**Comment by tde-banana-odoo** [comment]
@robodoo r+
